File: windows-app/Plantoir.Tests/SectionPublishStateTests.cs
using Plantoir.Core.Models;
using Xunit;

namespace Plantoir.Tests;

/// <summary>
/// Mirrors the mac's SectionPublishStateTests.swift — the " — Edited"
/// title-bar marker. Reads the same authored case list the mac suite runs,
/// from <c>contracts/app-rules.json</c> → <c>publishedFreshness</c>, so
/// both platforms are pinned against one specification rather than two
/// hand-typed copies of it. See "The ' — Edited' marker" in
/// documentation/05-build-pipeline.md.
/// </summary>
public class SectionPublishStateTests
{
    private static string NewCourseDirectory()
    {
        string path = Directory.CreateTempSubdirectory("section-publish-state-tests").FullName;
        return path;
    }

    // ---- The marker's wording -----------------------------------------

    [Fact]
    public void Marker_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("app-rules.json");
        var cases = doc["publishedFreshness"]!["marker"]!["cases"]!.AsArray();

        foreach (var entry in cases)
        {
            // A case flagged "in a sentence naming the section" tests a
            // DIFFERENT rule — that sentences like "Deploying ICS3U-S1" use
            // the bare name and never call WindowTitle at all. Covered by
            // TitleText's own callers, not by this method.
            if (entry!["inASentenceNamingTheSection"]?.GetValue<bool>() == true) continue;

            string @base = entry["base"]!.ToString();
            bool hasEdits = entry["hasUnpublishedEdits"]!.GetValue<bool>();
            string expected = entry["expectTitle"]!.ToString();
            Assert.Equal(expected, SectionPublishState.WindowTitle(@base, hasEdits));
        }
    }

    [Fact]
    public void Marker_Suffix_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("app-rules.json");
        string suffix = doc["publishedFreshness"]!["marker"]!["suffix"]!.ToString();
        Assert.Equal(SectionPublishState.WindowTitle("X", true), "X" + suffix);
    }

    // ---- Which files count ---------------------------------------------

    [Fact]
    public void FilesCounted_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("app-rules.json");
        var cases = doc["publishedFreshness"]!["filesCounted"]!["cases"]!.AsArray();

        foreach (var entry in cases)
        {
            string relativePath = entry!["path"]!.ToString().Replace('\\', '/');
            bool expectedCounts = entry["counts"]!.GetValue<bool>();

            // section1's own fingerprint is what every case in the contract
            // is written against.
            bool counts = SectionPublishState.CountsTowardFingerprint(relativePath, sectionNumber: 1);
            Assert.True(counts == expectedCounts, $"'{relativePath}' expected counts={expectedCounts}, got {counts}");
        }
    }

    [Fact]
    public void IsSectionFolderName_OnlyMatchesSectionPlusDigits()
    {
        Assert.True(SectionPublishState.IsSectionFolderName("section3"));
        Assert.True(SectionPublishState.IsSectionFolderName("section12"));
        Assert.False(SectionPublishState.IsSectionFolderName("sections"));
        Assert.False(SectionPublishState.IsSectionFolderName("section"));
        Assert.False(SectionPublishState.IsSectionFolderName("section3b"));
    }

    // ---- Self-publishing exclusion --------------------------------------

    [Fact]
    public void SelfPublishing_MatchesContract()
    {
        var doc = ContractLoader.LoadJson("app-rules.json");
        var cases = doc["publishedFreshness"]!["selfPublishing"]!["cases"]!.AsArray();

        foreach (var entry in cases)
        {
            string courseDirectory = entry!["courseDirectory"]!.ToString();
            string destinationPath = entry["destinationPath"]!.ToString();
            string? expectExcluded = entry["expectExcluded"]?.GetValue<string?>();

            var destinations = new[] { new CourseConfiguration.DeployDestination("local_folder", destinationPath) };
            var excluded = SectionPublishState.SelfPublishingSubpaths(courseDirectory, destinations);

            if (expectExcluded is null)
                Assert.Empty(excluded);
            else
                Assert.Contains(expectExcluded, excluded);
        }
    }

    [Fact]
    public void SelfPublishing_ExclusionIsCaseInsensitive()
    {
        // NTFS is case-insensitive but case-preserving: a teacher's saved
        // destination path can differ in case from the folder's actual
        // on-disk name. Found by adversarial review — the mac's own
        // filesystem never raised the question, but this one does.
        var destinations = new[]
        {
            new CourseConfiguration.DeployDestination("local_folder", @"C:\w\courses\ICS3U\Site"),
        };
        var excluded = SectionPublishState.SelfPublishingSubpaths(@"C:\w\courses\ICS3U", destinations);
        Assert.True(SectionPublishState.IsExcluded("site", excluded));
        Assert.True(SectionPublishState.IsExcluded("site/index.html", excluded));
        Assert.True(SectionPublishState.IsExcluded("SITE/index.html", excluded));
    }

    // ---- End-to-end: on disk, with real files ---------------------------

    [Fact]
    public void NeverPublished_ShowsNoEdits()
    {
        string course = NewCourseDirectory();
        File.WriteAllText(Path.Combine(course, "course_config.json"), "{}");
        Assert.False(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void UnchangedSincePublish_ShowsNoEdits()
    {
        string course = NewCourseDirectory();
        Directory.CreateDirectory(Path.Combine(course, "section1", "Classes"));
        File.WriteAllText(Path.Combine(course, "section1", "Classes", "Day 1.md"), "hello");

        string fingerprint = SectionPublishState.Fingerprint(course, 1);
        Assert.True(SectionPublishState.RecordPublish(course, 1, fingerprint, new[] { "netlify" }));

        Assert.False(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void EditingAPageAfterPublish_ShowsEdits()
    {
        string course = NewCourseDirectory();
        string page = Path.Combine(course, "section1", "Classes", "Day 1.md");
        Directory.CreateDirectory(Path.GetDirectoryName(page)!);
        File.WriteAllText(page, "hello");

        string fingerprint = SectionPublishState.Fingerprint(course, 1);
        SectionPublishState.RecordPublish(course, 1, fingerprint, new[] { "netlify" });

        Thread.Sleep(20);
        File.WriteAllText(page, "hello, edited");
        Assert.True(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void DeletingAPageAfterPublish_ShowsEdits()
    {
        string course = NewCourseDirectory();
        string page = Path.Combine(course, "section1", "Classes", "Day 1.md");
        Directory.CreateDirectory(Path.GetDirectoryName(page)!);
        File.WriteAllText(page, "hello");

        string fingerprint = SectionPublishState.Fingerprint(course, 1);
        SectionPublishState.RecordPublish(course, 1, fingerprint, new[] { "netlify" });

        File.Delete(page);
        Assert.True(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void AnotherSectionsPageChanging_DoesNotShowEdits()
    {
        string course = NewCourseDirectory();
        string ownPage = Path.Combine(course, "section1", "Classes", "Day 1.md");
        string otherPage = Path.Combine(course, "section2", "Classes", "Day 1.md");
        Directory.CreateDirectory(Path.GetDirectoryName(ownPage)!);
        Directory.CreateDirectory(Path.GetDirectoryName(otherPage)!);
        File.WriteAllText(ownPage, "hello");
        File.WriteAllText(otherPage, "hello");

        string fingerprint = SectionPublishState.Fingerprint(course, 1);
        SectionPublishState.RecordPublish(course, 1, fingerprint, new[] { "netlify" });

        Thread.Sleep(20);
        File.WriteAllText(otherPage, "hello, edited");
        Assert.False(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void SharedCourseLevelPageChanging_ShowsEditsForEverySection()
    {
        string course = NewCourseDirectory();
        string sharedPage = Path.Combine(course, "Unit 1", "Lesson.md");
        Directory.CreateDirectory(Path.GetDirectoryName(sharedPage)!);
        File.WriteAllText(sharedPage, "hello");

        string fp1 = SectionPublishState.Fingerprint(course, 1);
        string fp2 = SectionPublishState.Fingerprint(course, 2);
        SectionPublishState.RecordPublish(course, 1, fp1, new[] { "netlify" });
        SectionPublishState.RecordPublish(course, 2, fp2, new[] { "netlify" });

        Thread.Sleep(20);
        File.WriteAllText(sharedPage, "hello, edited");
        Assert.True(SectionPublishState.HasUnpublishedEdits(course, 1));
        Assert.True(SectionPublishState.HasUnpublishedEdits(course, 2));
    }

    [Fact]
    public void UnreadableStamp_ShowsNoEdits()
    {
        string course = NewCourseDirectory();
        Directory.CreateDirectory(Path.Combine(course, ".publish_state"));
        File.WriteAllText(SectionPublishState.StampPath(course, 1), "not json at all {{{");
        Assert.False(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void StampFile_IsHiddenFromItsOwnFingerprint()
    {
        // Load-bearing: counted, every publish would end by declaring the
        // section edited.
        string course = NewCourseDirectory();
        Directory.CreateDirectory(Path.Combine(course, "section1"));
        File.WriteAllText(Path.Combine(course, "section1", "page.md"), "hello");

        string fingerprint = SectionPublishState.Fingerprint(course, 1);
        SectionPublishState.RecordPublish(course, 1, fingerprint, new[] { "netlify" });

        // Recomputing right after the stamp was written must still agree —
        // if .publish_state counted, this would immediately read as edited.
        Assert.False(SectionPublishState.HasUnpublishedEdits(course, 1));
    }

    [Fact]
    public void SymlinkedFolder_ContributesItsContents()
    {
        string course = NewCourseDirectory();
        string vault = Directory.CreateTempSubdirectory("section-publish-vault").FullName;
        File.WriteAllText(Path.Combine(vault, "diagram.png"), "pretend image bytes");

        string mediaLink = Path.Combine(course, "Media");
        try { Directory.CreateSymbolicLink(mediaLink, vault); }
        catch (IOException) { return; } // no privilege to create symlinks on this machine — skip rather than fail spuriously

        string fingerprint = SectionPublishState.Fingerprint(course, 1);
        SectionPublishState.RecordPublish(course, 1, fingerprint, new[] { "netlify" });
        Assert.False(SectionPublishState.HasUnpublishedEdits(course, 1));

        Thread.Sleep(20);
        File.WriteAllText(Path.Combine(vault, "diagram.png"), "pretend image bytes, edited");
        Assert.True(SectionPublishState.HasUnpublishedEdits(course, 1));
    }
}

File: mac-app/Tests/QuartzTeachersTests/SectionPublishStateTests.swift
import XCTest
@testable import QuartzTeachers

/// Runs the `publishedFreshness` half of `contracts/app-rules.json` — the
/// " — Edited" marker a section window shows when its pages have changed
/// since it was last published.
///
/// The cases are data so the Windows suite runs the identical list; the
/// setup is Swift because only this side can make a course folder on disk.
/// Every `when` in the contract is answered by driving the REAL check
/// against a real folder rather than by asserting a helper in isolation —
/// the failure this feature must never have (telling a teacher a page went
/// out when it did not) lives in the walk, not in the arithmetic.
final class SectionPublishStateTests: XCTestCase {

    // MARK: - Stored properties

    private var courseDirectory: URL!

    // MARK: - Set-up

    override func setUpWithError() throws {
        courseDirectory = FileManager.default.temporaryDirectory
            .appendingPathComponent("publish-state-\(UUID().uuidString)")
            .appendingPathComponent("ICS3U")
        try write("{\"course_code\": \"ICS3U\"}", to: "course_config.json")
        try write("# Day one", to: "section1/Classes/Unit 1, Day 1.md")
        try write("# Day one, section two", to: "section2/Classes/Unit 1, Day 1.md")
        try write("# A shared lesson", to: "Unit 1/Lesson.md")
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: courseDirectory.deletingLastPathComponent())
    }

    // MARK: - The marker itself

    func testTitleCarriesTheMarkerTheContractNames() throws {
        let marker: [String: Any] = try XCTUnwrap(
            try SectionPublishStateTests.rules()["marker"] as? [String: Any]
        )
        for testCase in try XCTUnwrap(marker["cases"] as? [[String: Any]]) {
            let base: String = try XCTUnwrap(testCase["base"] as? String)
            let edited: Bool = try XCTUnwrap(testCase["hasUnpublishedEdits"] as? Bool)
            let expected: String = try XCTUnwrap(testCase["expectTitle"] as? String)

            // A sentence NAMING the section carries the bare name — the
            // marker belongs to the title bar and nowhere else.
            if testCase["inASentenceNamingTheSection"] as? Bool == true {
                XCTAssertEqual(base, expected)
                continue
            }
            XCTAssertEqual(
                SectionPublishState.windowTitle(base: base, hasUnpublishedEdits: edited),
                expected,
                "The title bar must say exactly what the contract says it says"
            )
        }
    }

    /// Reported from a real deploy: the progress panel said "Deploying
    /// ICS3U-S1 — Edited", which reads as though "Edited" were something
    /// being published. Every sentence that names the section is passed
    /// the name, never the title.
    func testASentenceNamingTheSectionNeverCarriesTheMarker() {
        let deploying: String = SectionDetailView.deployProgressTitle(
            sectionName: "ICS3U-S1", isRunning: true, legCount: 2,
            currentDestinationDescription: "Netlify"
        )
        XCTAssertFalse(deploying.contains("Edited"))
        XCTAssertTrue(deploying.contains("ICS3U-S1"))

        let previewing: String = SectionDetailView.previewTaskTitle(
            isPreparing: true, sectionName: "ICS3U-S1"
        )
        XCTAssertFalse(previewing.contains("Edited"))
        XCTAssertTrue(previewing.contains("ICS3U-S1"))
    }

    // MARK: - Which files count

    /// Driven through the REAL walk — every path in the contract is
    /// written to disk, changed, and the marker asked about — rather than
    /// through the filter function alone.
    ///
    /// The difference is not academic. Reviewed adversarially, the filter
    /// said `Media/diagram.png` counted, the contract said it counted, and
    /// the walk skipped the whole folder whenever `Media` was a symlink,
    /// which is how build_site.py itself sets it up. A test that asks the
    /// filter agrees with a bug; a test that asks the walk finds it.
    func testTheRightFilesCountTowardTheFingerprint() throws {
        let section: [String: Any] = try XCTUnwrap(
            try SectionPublishStateTests.rules()["filesCounted"] as? [String: Any]
        )
        for testCase in try XCTUnwrap(section["cases"] as? [[String: Any]]) {
            let path: String = try XCTUnwrap(testCase["path"] as? String)
            let counts: Bool = try XCTUnwrap(testCase["counts"] as? Bool)
            let why: String = testCase["why"] as? String ?? ""

            try write("first", to: path)
            let before: String = SectionPublishState.fingerprint(
                courseDirectory: courseDirectory, sectionNumber: 1
            )
            try write("second, and longer", to: path)
            let after: String = SectionPublishState.fingerprint(
                courseDirectory: courseDirectory, sectionNumber: 1
            )

            XCTAssertEqual(
                before != after, counts,
                "\(path) — the WALK must agree with the contract, not just the filter. \(why)"
            )
            XCTAssertEqual(
                SectionPublishState.countsTowardFingerprint(relativePath: path, sectionNumber: 1),
                counts,
                "\(path): \(why)"
            )
        }
    }

    /// A symlinked page, and a symlinked folder, are content the site is
    /// built from. Neither is a regular file, and `FileManager`'s
    /// enumerator follows neither — so both were invisible until this
    /// test existed.
    func testSymlinkedContentIsSeen() throws {
        let outside: URL = courseDirectory
            .deletingLastPathComponent()
            .appendingPathComponent("vault")
        try FileManager.default.createDirectory(
            at: outside.appendingPathComponent("Media"), withIntermediateDirectories: true
        )
        try "diagram".write(
            to: outside.appendingPathComponent("Media/diagram.svg"),
            atomically: true, encoding: .utf8
        )
        try "linked page".write(
            to: outside.appendingPathComponent("Linked.md"), atomically: true, encoding: .utf8
        )
        try FileManager.default.createSymbolicLink(
            at: courseDirectory.appendingPathComponent("Media"),
            withDestinationURL: outside.appendingPathComponent("Media")
        )
        try FileManager.default.createSymbolicLink(
            at: courseDirectory.appendingPathComponent("Unit 1/Linked.md"),
            withDestinationURL: outside.appendingPathComponent("Linked.md")
        )

        publishNow()
        XCTAssertFalse(isEdited())

        try "a completely different diagram".write(
            to: outside.appendingPathComponent("Media/diagram.svg"),
            atomically: true, encoding: .utf8
        )
        XCTAssertTrue(isEdited(), "A change inside a symlinked Media folder is a change to the site")

        publishNow()
        try "a completely different linked page".write(
            to: outside.appendingPathComponent("Linked.md"), atomically: true, encoding: .utf8
        )
        XCTAssertTrue(isEdited(), "A symlinked page is still a page")
    }

    /// A course that publishes into a folder inside itself must not feed
    /// its own marker — `deploy.py` writes the whole site there.
    func testACourseThatPublishesIntoItselfIsNotItsOwnEdit() throws {
        let excluded: [String] = SectionPublishState.selfPublishingSubpaths(
            courseDirectory: courseDirectory,
            destinations: [
                CourseConfiguration.DeployDestination(
                    type: "local_folder", path: courseDirectory.appendingPathComponent("site").path
                ),
            ]
        )
        XCTAssertEqual(excluded, ["site"])

        let before: String = SectionPublishState.fingerprint(
            courseDirectory: courseDirectory, sectionNumber: 1, excludingRelativePaths: excluded
        )
        try write("<html>the whole built site</html>", to: "site/index.html")
        XCTAssertEqual(
            SectionPublishState.fingerprint(
                courseDirectory: courseDirectory, sectionNumber: 1, excludingRelativePaths: excluded
            ),
            before,
            "Publishing into the course folder must not read as editing it"
        )
    }

    /// A destination somewhere else on the disk excludes nothing.
    func testAnOrdinaryPublishFolderExcludesNothing() {
        XCTAssertEqual(
            SectionPublishState.selfPublishingSubpaths(
                courseDirectory: courseDirectory,
                destinations: [
                    CourseConfiguration.DeployDestination(
                        type: "local_folder", path: "/Users/teacher/Sites/ICS3U"
                    ),
                ]
            ),
            []
        )
    }

    /// `section3` is a section folder; `sections` and `section3b` are two
    /// folders a teacher is entirely free to make, and neither belongs to
    /// anybody. Getting this wrong would drop a real folder of pages out of
    /// every section's fingerprint.
    func testOnlyRealSectionFoldersAreTreatedAsOne() {
        XCTAssertTrue(SectionPublishState.isSectionFolderName("section3"))
        XCTAssertTrue(SectionPublishState.isSectionFolderName("section12"))
        XCTAssertFalse(SectionPublishState.isSectionFolderName("sections"))
        XCTAssertFalse(SectionPublishState.isSectionFolderName("section"))
        XCTAssertFalse(SectionPublishState.isSectionFolderName("section3b"))
        XCTAssertFalse(SectionPublishState.isSectionFolderName("Section 3"))
    }

    // MARK: - When the marker is shown

    func testEveryCaseTheContractNames() throws {
        var expectations: [String: Bool] = [:]
        for rule in try XCTUnwrap(
            try SectionPublishStateTests.rules()["whenShown"] as? [[String: Any]]
        ) {
            expectations[try XCTUnwrap(rule["when"] as? String)] =
                try XCTUnwrap(rule["expectEdited"] as? Bool)
        }

        // Never published.
        XCTAssertEqual(
            isEdited(), expectations["the section has never been published"],
            "A course nobody has published yet has not been edited since anything"
        )

        // Nothing changed since the last publish.
        publishNow()
        XCTAssertEqual(
            isEdited(), expectations["nothing has changed since the last publish"]
        )

        // A page this section alone uses.
        try write("# Day one, rewritten this morning", to: "section1/Classes/Unit 1, Day 1.md")
        XCTAssertEqual(
            isEdited(), expectations["a page this section alone uses has changed"]
        )

        // A page shared with the other sections.
        publishNow()
        try write("# A shared lesson, rewritten", to: "Unit 1/Lesson.md")
        XCTAssertEqual(
            isEdited(),
            expectations["a page this section SHARES with other sections has changed"]
        )

        // A page removed.
        publishNow()
        try FileManager.default.removeItem(
            at: courseDirectory.appendingPathComponent("Unit 1/Lesson.md")
        )
        XCTAssertEqual(
            isEdited(), expectations["a page has been DELETED since the last publish"],
            "A fingerprint of the file inventory is what makes a deletion visible"
        )

        // The configuration.
        publishNow()
        try write("{\"course_code\": \"ICS3U\", \"font\": \"Inter\"}", to: "course_config.json")
        XCTAssertEqual(
            isEdited(), expectations["course_config.json has changed"]
        )

        // An unreadable stamp.
        publishNow()
        try write("this is not JSON", to: ".publish_state/section1.json")
        XCTAssertEqual(
            isEdited(), expectations["the stamp file cannot be read"],
            "An unreadable stamp must never be able to claim a section is up to date"
        )
    }

    /// The other section's pages are not this section's site. Editing
    /// section 2 must leave section 1's window alone — the case that would
    /// make the marker meaningless in the courses that have most sections.
    func testAnotherSectionsPagesDoNotMarkThisOne() throws {
        publishNow()
        try write("# Rewritten for the other class", to: "section2/Classes/Unit 1, Day 1.md")
        XCTAssertFalse(isEdited())
    }

    /// The stamp lives in a hidden folder for exactly this reason: writing
    /// it must not itself count as an edit, or every publish would end by
    /// declaring the section edited.
    func testRecordingAPublishDoesNotItselfCountAsAnEdit() {
        publishNow()
        XCTAssertFalse(isEdited())
        publishNow()
        XCTAssertFalse(isEdited())
    }

    /// The stamp says where the pages went, because a course since pointed
    /// somewhere else has never published THERE.
    func testTheStampRecordsWhereItWent() {
        publishNow(destinations: ["netlify", "cloudflare_pages"])
        let stamp = SectionPublishState.stamp(courseDirectory: courseDirectory, sectionNumber: 1)
        XCTAssertEqual(stamp?.destinations, ["netlify", "cloudflare_pages"])
    }

    // MARK: - Helpers

    private func isEdited() -> Bool {
        return SectionPublishState.hasUnpublishedEdits(
            courseDirectory: courseDirectory,
            sectionNumber: 1
        )
    }

    private func publishNow(destinations: [String] = ["netlify"]) {
        SectionPublishState.recordPublish(
            courseDirectory: courseDirectory,
            sectionNumber: 1,
            fingerprint: SectionPublishState.fingerprint(
                courseDirectory: courseDirectory,
                sectionNumber: 1
            ),
            destinations: destinations
        )
    }

    private func write(_ text: String, to relativePath: String) throws {
        let url: URL = courseDirectory.appendingPathComponent(relativePath)
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try text.write(to: url, atomically: true, encoding: .utf8)
    }

    private static func rules() throws -> [String: Any] {
        let contracts: URL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("contracts")
            .appendingPathComponent("app-rules.json")
        let whole: [String: Any] = try XCTUnwrap(
            try JSONSerialization.jsonObject(with: try Data(contentsOf: contracts)) as? [String: Any]
        )
        return try XCTUnwrap(whole["publishedFreshness"] as? [String: Any])
    }
}

/// The other half of `publishedFreshness` — when a publish is RECORDED at
/// all. Every rule in `whenRecorded` was authored and pinned by nothing
/// until an adversarial review pointed out that a quarter of the contract
/// was decoration, which is how the fingerprint came to be taken at a
/// moment the contract said it was not.
@MainActor
final class SectionPublishRecordingTests: XCTestCase {

    // MARK: - Stored properties

    private var course: Course!

    // MARK: - Set-up

    override func setUpWithError() throws {
        let directory: URL = FileManager.default.temporaryDirectory
            .appendingPathComponent("publish-record-\(UUID().uuidString)")
            .appendingPathComponent("ICS3U")
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        course = Course(
            code: "ICS3U",
            directoryURL: directory,
            configuration: CourseConfiguration(values: ["course_code": "ICS3U"], lastSavedData: Data())
        )
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: course.directoryURL.deletingLastPathComponent())
    }

    // MARK: - Functions

    func testEveryRecordingRuleTheContractNames() throws {
        let netlify: CourseConfiguration.DeployDestination =
            CourseConfiguration.DeployDestination(type: "netlify", path: "")
        let cloudflare: CourseConfiguration.DeployDestination =
            CourseConfiguration.DeployDestination(type: "cloudflare_pages", path: "")

        XCTAssertTrue(
            recorded(legs: [finished(netlify, succeeded: true)]),
            "every configured destination succeeded"
        )
        XCTAssertTrue(
            recorded(legs: [finished(netlify, succeeded: true), finished(cloudflare, succeeded: true)])
        )
        XCTAssertFalse(
            recorded(legs: [finished(netlify, succeeded: true), finished(cloudflare, succeeded: false)]),
            "one of two destinations failed — the section has NOT published"
        )
        XCTAssertFalse(
            recorded(legs: [finished(netlify, succeeded: false)]),
            "the publish was cancelled or stopped"
        )
        XCTAssertFalse(
            recorded(legs: [finished(netlify, succeeded: false, buildFailed: true)]),
            "the shared build failed"
        )
        XCTAssertFalse(recorded(legs: []), "no destination ran at all")
    }

    /// The fingerprint must be taken before the BUILD, not after it. The
    /// build is the longest part of a publish, so a page edited while it
    /// runs must leave the marker up rather than be stamped as published.
    func testAPageEditedDuringTheBuildIsNotMarkedAsPublished() throws {
        let page: URL = course.directoryURL.appendingPathComponent("section1/Classes/Day.md")
        try FileManager.default.createDirectory(
            at: page.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try "before".write(to: page, atomically: true, encoding: .utf8)

        // What the runner does at the top of its loop, before the build.
        let atStart: String = SectionPublishState.fingerprint(
            courseDirectory: course.directoryURL, sectionNumber: 1
        )
        // …the build runs, and the teacher edits a page while it does.
        try "edited while the build was running".write(to: page, atomically: true, encoding: .utf8)

        let runner: MultiDestinationDeployRunner = MultiDestinationDeployRunner()
        runner.legs = [
            finished(CourseConfiguration.DeployDestination(type: "netlify", path: ""), succeeded: true),
        ]
        runner.recordWhatWentOut(course: course, sectionNumber: 1, fingerprint: atStart)

        XCTAssertTrue(
            SectionPublishState.hasUnpublishedEdits(
                courseDirectory: course.directoryURL, sectionNumber: 1
            ),
            "An edit made during the publish did not go out, and must keep the marker up"
        )
    }

    /// A scheduled deploy never goes through the deploy runner — launchd
    /// runs a shell script — so it needs its own path to the same record.
    func testAScheduledDeployRecordsWhatWentOut() throws {
        let sentinel: URL = ScheduledDeploy.successSentinelURL(courseCode: "ICS3U", sectionNumber: 1)
        try FileManager.default.createDirectory(
            at: sentinel.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        defer { try? FileManager.default.removeItem(at: sentinel) }

        let section: (courseDirectory: URL, courseCode: String, sectionNumber: Int) =
            (course.directoryURL, "ICS3U", 1)

        // No sentinel: the script did not say every destination worked.
        ScheduledDeploy.recordScheduledPublish(section: section, fingerprint: "abc")
        XCTAssertNil(SectionPublishState.stamp(courseDirectory: course.directoryURL, sectionNumber: 1))

        try "netlify cloudflare_pages\n".write(to: sentinel, atomically: true, encoding: .utf8)
        ScheduledDeploy.recordScheduledPublish(section: section, fingerprint: "abc")
        let stamp = SectionPublishState.stamp(courseDirectory: course.directoryURL, sectionNumber: 1)
        XCTAssertEqual(stamp?.fingerprint, "abc")
        XCTAssertEqual(stamp?.destinations, ["netlify", "cloudflare_pages"])
        XCTAssertFalse(
            FileManager.default.fileExists(atPath: sentinel.path),
            "The sentinel is consumed, or tonight's failure reads as tomorrow's success"
        )
    }

    /// The agent has to tell the app which section it just published, or
    /// the app cannot record anything.
    func testTheScheduledScriptAndAgentCarryTheSection() {
        let workspace: URL = URL(fileURLWithPath: "/Users/teacher/Teaching")
        let named = ScheduledDeploy.requestedSection(from: [
            "/Applications/Plantoir.app/Contents/MacOS/Plantoir",
            ScheduledDeploy.runFlag, "/tmp/script.sh",
            ScheduledDeploy.sectionFlag, workspace.path, "ICS3U", "2",
        ])
        XCTAssertEqual(named?.courseCode, "ICS3U")
        XCTAssertEqual(named?.sectionNumber, 2)
        XCTAssertEqual(
            named?.courseDirectory.path,
            workspace.appendingPathComponent("courses/ICS3U").path
        )
        XCTAssertNil(ScheduledDeploy.requestedSection(from: ["Plantoir"]))

        let script: String = ScheduledDeploy.oneShotCommand(
            courseCode: "ICS3U",
            sectionNumber: 1,
            workspaceURL: workspace,
            deployArgumentsList: [["ICS3U", "1"], ["ICS3U", "1", "--target", "cloudflare"]],
            destinationTypes: ["netlify", "cloudflare_pages"]
        )
        XCTAssertTrue(
            script.contains("ALL_OK=0"),
            "Each destination's own result must be tracked, not just the last one's"
        )
        XCTAssertTrue(script.contains("netlify cloudflare_pages"))
        XCTAssertTrue(
            script.contains(ScheduledDeploy.successSentinelURL(courseCode: "ICS3U", sectionNumber: 1).path)
        )
    }

    // MARK: - Helpers

    private func finished(
        _ destination: CourseConfiguration.DeployDestination,
        succeeded: Bool,
        buildFailed: Bool = false
    ) -> MultiDestinationDeployRunner.Leg {
        var leg: MultiDestinationDeployRunner.Leg = MultiDestinationDeployRunner.Leg(
            destination: destination
        )
        leg.isFinished = true
        leg.succeeded = succeeded
        leg.buildFailed = buildFailed
        return leg
    }

    private func recorded(legs: [MultiDestinationDeployRunner.Leg]) -> Bool {
        try? FileManager.default.removeItem(
            at: SectionPublishState.stampURL(courseDirectory: course.directoryURL, sectionNumber: 1)
        )
        let runner: MultiDestinationDeployRunner = MultiDestinationDeployRunner()
        runner.legs = legs
        runner.recordWhatWentOut(course: course, sectionNumber: 1, fingerprint: "fingerprint")
        return SectionPublishState.stamp(
            courseDirectory: course.directoryURL, sectionNumber: 1
        ) != nil
    }
}
